linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala

```diff
@@ -1557,41 +1744,60 @@ private[emitter] object CoreJSLib {
       }
     }
 
-    private def defineAsArrayOfPrimitiveFunctions(): Unit = {
+    private def defineSpecializedAsArrayOfFunctions(): Unit = {
       if (asInstanceOfs != CheckedBehavior.Unchecked) {
-        for (primRef <- orderedPrimRefsWithoutVoid) {
+        for (typeRef <- specializedArrayTypeRefs) {
+          val encodedName = typeRef match {
+            case typeRef: PrimRef => typeRef.charCode.toString()
+            case _                => "L" + ObjectClass.nameString + ";"
+          }
+
           val obj = varRef("obj")
           val depth = varRef("depth")
-          buf += extractWithGlobals(globalFunctionDef("asArrayOf", primRef, paramList(obj, depth), {
-            If(Apply(typeRefVar("isArrayOf", primRef), obj :: depth :: Nil) || (obj === Null()), {
+          buf += extractWithGlobals(globalFunctionDef("asArrayOf", typeRef, paramList(obj, depth), {
+            If(Apply(globalVar("isArrayOf", typeRef), obj :: depth :: Nil) || (obj === Null()), {
               Return(obj)
             }, {
-              genCallHelper("throwArrayCastException", obj,
-                  str(primRef.charCode.toString()), depth)
+              genCallHelper("throwArrayCastException", obj, str(encodedName), depth)
             })
           }))
         }
       }
     }
 
-    private def definePrimitiveTypeDatas(): Unit = {
-      for {
-        (primRef, zero) <- List(
-            (VoidRef, Undefined()),
-            (BooleanRef, bool(false)),
-            (CharRef, int(0)),
-            (ByteRef, int(0)),
-            (ShortRef, int(0)),
-            (IntRef, int(0)),
-            (LongRef, if (allowBigIntsForLongs) genLongZero() else str("longZero")),
-            (FloatRef, double(0)),
-            (DoubleRef, double(0))
-        )
-      } {
+    private def defineSpecializedTypeDatas(): Unit = {
+      // d_O must be first to correctly populate the parentData of array classes
+      locally {
+        val typeData = Apply(
+            New(globalVar("TypeData", CoreVar), Nil) DOT "initObjectClass", Nil)
+
+        buf += extractWithGlobals(globalVarDef("d", ObjectClass, typeData))
+        buf += {
+          globalVar("c", ObjectClass).prototype DOT "$classData" :=
+            globalVar("d", ObjectClass)
+        }
+      }
+
+      for (primRef <- orderedPrimRefs) {
+        /* Zero value, for use by the intrinsified code of
+         * `scala.collection.mutable.ArrayBuilder.genericArrayBuilderResult`.
+         * This code is Scala-specific, and "unboxes" `null` as the zero of
+         * primitive types. For `void`, it is even more special, as it produces
+         * a boxed Unit value, which is `undefined` (although `VoidRef`/`NoType`
+         * doesn't have a zero value per se).
+         */
```

Comment:
I do not understand this explanation: With the recent change of `makeNativeArrayWrapper`, `zero` doesn't even seem to be needed anymore. We used to need it to initialize arrays properly, but since all non-object arrays are specialized, they have a static reference to their `zero`.

Reply:
`zero` is still used by the translation of `"zeroOf"` in `FunctionEmitter`. `"zeroOf"` is generated by the optimizer as the intrinsic for `ArrayBuilder.zeroOf`.